cwms-data-api/src/main/java/cwms/cda/data/dao/AuthDao.java

```diff
@@ -0,0 +1,223 @@
+                }
+            } catch (SQLException ex) {
+                logger.atWarning().withCause(ex).log("Failed API key check");
+                throw new CwmsAuthException("Authorized failed",ex);
+            }
+        });
+    }
+
+    /**
+     * Retrieve roles a user has. 
+     * @param user
+     * @return
+     */
+    private Set<RouteRole> getRolesForUser(String user) {
+        final Set<RouteRole> roles = new HashSet<>();
+        dsl.connection(c->{
+            setSessionForAuthCheck(c);
+            try (PreparedStatement getRoles = c.prepareStatement(RETRIEVE_GROUPS_OF_USER);) {
+                getRoles.setString(1,user);
+                try (ResultSet rs = getRoles.executeQuery()) {
+                    while (rs.next()) {
+                        roles.add(new Role(rs.getString(1)));
+                    }
+                }
+            } catch (SQLException ex) {
+                logger.atWarning().withCause(ex).log("Failed to retrieve roles for user.");
+            }
+        });
+        return roles;
+    }
+
+    /**
+     * Allows connection to be correctly setup, key is used to assert user, then the office
+     * is set to the user specified office for further checks within the database.
+     *
+     * NOTE: side affect on ctx for this session.
+     * 
+     * @param ctx javalin context if additional parameters are required.
+     * @param user username, which is ignored except a log message
+     * @param key the API key that was presented for this connection
+     */
+    public static void prepareContextWithUser(Context ctx, DataApiPrincipal p) throws SQLException {
+        Objects.requireNonNull(ctx, "A valid Javalin Context must be provided to this call.");
+        Objects.requireNonNull(p, "A valid data api principal must be provided to this call.");
+        logger.atInfo().log("Validated Api Key for user=%s", p.getName());
+        DataSource dataSource = ctx.attribute(ApiServlet.DATA_SOURCE);
+        ConnectionPreparer userPreparer = new DirectUserPreparer(p.getName());
+
+        if (dataSource instanceof ConnectionPreparingDataSource) {
+            ConnectionPreparingDataSource cpDs = (ConnectionPreparingDataSource)dataSource;
+            ConnectionPreparer existingPreparer = cpDs.getPreparer();
+
+            // Have it do our extra step last.
+            cpDs.setPreparer(new DelegatingConnectionPreparer(existingPreparer, userPreparer));
+        } else {
+            ctx.attribute(ApiServlet.DATA_SOURCE,
+                          new ConnectionPreparingDataSource(userPreparer, dataSource));
+        }
+    }
+
+    /**
+     * Primary logic to determine if a given principal can perform the desired operation.
+     * Throws exception if not valid, otherwise just returns.
+     * @param ctx
+     * @param p
+     * @param routeRoles
+     */
+    public static void isAuthorized(Context ctx, Optional<DataApiPrincipal> p, Set<RouteRole> routeRoles) throws CwmsAuthException {    
```

Comment:
Anti-pattern using an Optional parameter

Reply:
Nulls, Optionals , and boolean flag parameters all fall under the same antipattern. It winds up having the function caller needing to know that there are essentially different implementations of the same method that route depending on which parameters were used. This becomes a maintenance issue in public API's as you then have the function callers dictating the implementation support rather than having the API's dictate what they support. 

https://medium.com/@amlcurran/clean-code-the-curse-of-a-boolean-parameter-c237a830b7a3

Reply:
Okay, I can buy that. Having an actual function name could often be more readable.